🌱 bump gosec to 2.17.0, and fix gosec for submodules

tuminoid opened the pull request:
Gosec does not handle submodules well. Ignore them and run them separately so they run against correct dependencies.

Bump gosec to 2.17.0, and fix new errors it has discovered.
Remove G107 exclude, it is no longer present.

NOTE: In CI, the gosec image is coming via project-infra, gosec image change is only affecting local runs for now.

tuminoid commented:
/test all

tuminoid commented:
Locally executing gives this output (before fixing the issues), but notably failing on dependencies as well:

```console
+ gosec -severity medium --confidence medium -quiet -concurrency 8 '-exclude-dir=api' '-exclude-dir=test' ./...
+ gosec -severity medium --confidence medium -quiet ./api/...
+ gosec -severity medium --confidence medium -quiet ./test/...
Results:

Golang errors in file: [/workdir/test/e2e/cert_rotation.go]:

  > [line 18 : column 2] - could not import sigs.k8s.io/cluster-api/test/framework (invalid package name: "")

  > [line 19 : column 2] - could not import sigs.k8s.io/cluster-api/test/framework/clusterctl (invalid package name: "")


Golang errors in file: [/workdir/test/e2e/common.go]:

  > [line 25 : column 2] - could not import golang.org/x/crypto/ssh (invalid package name: "")


Golang errors in file: [/workdir/test/e2e/pivoting.go]:

  > [line 21 : column 14] - could not import github.com/docker/docker/api/types (invalid package name: "")

  > [line 22 : column 17] - could not import github.com/docker/docker/api/types/container (invalid package name: "")

  > [line 23 : column 9] - could not import github.com/docker/docker/client (invalid package name: "")



[/workdir/test/e2e/common.go:633] - G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
    632: 	home := os.Getenv("HOME")
  > 633: 	privkey, err := os.ReadFile(path.Join(home, "/.ssh/id_rsa"))
    634: 	if err != nil {



[/workdir/test/e2e/common.go:178] - G107 (CWE-88): Potential HTTP request made with variable url (Confidence: MEDIUM, Severity: MEDIUM)
    177: 	// Get the data
  > 178: 	resp, err := http.Get(url) //nolint:noctx
    179: 	if err != nil {



[/workdir/test/e2e/pivoting.go:628] - G306 (CWE-276): Expect WriteFile permissions to be 0600 or less (Confidence: HIGH, Severity: MEDIUM)
    627: 		}
  > 628: 		writeErr := os.WriteFile(filepath.Join(logDir, "stdout.log"), out, 0444)
    629: 		Expect(writeErr).ToNot(HaveOccurred())



[/workdir/test/e2e/pivoting.go:624] - G306 (CWE-276): Expect WriteFile permissions to be 0600 or less (Confidence: HIGH, Severity: MEDIUM)
    623: 		if err != nil {
  > 624: 			writeErr := os.WriteFile(filepath.Join(logDir, "stderr.log"), []byte(err.Error()), 0444)
    625: 			Expect(writeErr).ToNot(HaveOccurred())



[/workdir/test/e2e/common.go:667] - G306 (CWE-276): Expect WriteFile permissions to be 0600 or less (Confidence: HIGH, Severity: MEDIUM)
    666: 	result := strings.TrimSuffix(stdoutBuf.String(), "\n") + "\n" + strings.TrimSuffix(stderrBuf.String(), "\n")
  > 667: 	if err := os.WriteFile(logFile, []byte(result), 0o777); err != nil {
    668: 		return fmt.Errorf("error writing log file: %w", err)



[/workdir/test/e2e/common.go:165] - G306 (CWE-276): Expect WriteFile permissions to be 0600 or less (Confidence: HIGH, Severity: MEDIUM)
    164: 		formattedSha256sum := fmt.Sprintf("%x", sha256sum)
  > 165: 		err = os.WriteFile(fmt.Sprintf("%s/%s.sha256sum", ironicImageDir, rawImageName), []byte(formattedSha256sum), 0544)
    166: 		Expect(err).To(BeNil())



Summary:
  Gosec  : dev
  Files  : 11
  Lines  : 3153
  Nosec  : 15
  Issues : 6

```

tuminoid commented:
Running `gosec` after `cd` seems to be working better than pointing it to subdirectory.

```sh
    gosec -severity medium --confidence medium -quiet \
        -concurrency "${GO_CONCURRENCY}" -exclude-dir=api -exclude-dir=test ./...
    (cd api && gosec -severity medium --confidence medium -quiet ./...)
    (cd test && gosec -severity medium --confidence medium -quiet ./...)
```

tuminoid commented:
/test-centos-e2e-integration-main

tuminoid commented:
/test-centos-e2e-integration-main
/test-centos-e2e-integration-main

tuminoid commented on test/e2e/common.go:
```diff
@@ -629,8 +629,12 @@ func MachineToIPAddress(ctx context.Context, cli client.Client, m *clusterv1.Mac
+		return fmt.Errorf("could not get home directory: %w", err)
+	}
+	key_path := path.Join(filepath.Clean(home), ".ssh", "id_rsa")
+	privkey, err := os.ReadFile(key_path) //#nosec
```
gosec is still giving G304 here. Variable no longer is tainted via `$HOME` and is cleaned to be safe. Any ideas besides `//#nosec`? @lentzi90

lentzi90 replied:
No idea. Is there even a situation where it would not complain unless a constant is used? :shrug:

tuminoid replied:
Apparently it wants constant, and no matter how much it is validated it is not going to be happy. Ignored it.

tuminoid commented on test/e2e/common.go:
```diff
@@ -175,7 +175,7 @@ func EnsureImage(k8sVersion string) (imageURL string, imageChecksum string) {
 func DownloadFile(filePath string, url string) error {
 	// Get the data
-	resp, err := http.Get(url) //nolint:noctx
+	resp, err := http.Get(url) //nolint:noctx #nosec
```
`#nosec` here is not doign anything (not like this nor as `//#nosec`, and if moved to start of comment, then `nolint` has no effect and golangci-lint fails. How to make both work on the same line?

lentzi90 replied:
We may have to put one of them above. Something like 
```go
/* #nosec */
resp, err := http.Get(url) //nolint:noctx
```

kashifest replied:
or //nolint:noctx,gosec

tuminoid replied:
Used @lentzi90's suggestion in the end.

tuminoid commented:
/test-centos-e2e-integration-main
/test-centos-e2e-integration-main

kashifest commented:
I was testing how to put this nolint and nosec together and I removed one but I didn't get any error

```
./hack/gosec.sh
+ IS_CONTAINER=false
+ CONTAINER_RUNTIME=docker
+ GO_CONCURRENCY=8
+ '[' false '!=' false ']'
+ docker run --rm --env IS_CONTAINER=TRUE --volume /Users/kashif/Documents/Work/cluster-api-provider-metal3:/workdir:ro,z --entrypoint sh --workdir /workdir docker.io/securego/gosec:2.17.0@sha256:4ea9b6053eac43abda841af5885bbd31ee1bf7289675545b8858bcedb40b4fa8 /workdir/hack/gosec.sh
+ IS_CONTAINER=TRUE
+ CONTAINER_RUNTIME=podman
+ GO_CONCURRENCY=8
+ '[' TRUE '!=' false ]
+ export 'XDG_CACHE_HOME=/tmp/.cache'
+ gosec -severity medium --confidence medium -quiet -concurrency 8 '-exclude-dir=api' '-exclude-dir=test' ./...
+ cd api
+ gosec -severity medium --confidence medium -quiet -concurrency 8 ./...
+ cd test
+ gosec '-exclude=G107' -severity medium --confidence medium -quiet -concurrency 8 ./...
kashif@Kashifs-MacBook-Pro cluster-api-provider-metal3 % git diff                       
diff --git a/test/e2e/common.go b/test/e2e/common.go
index 0c271d3a..9c9f218a 100644
--- a/test/e2e/common.go
+++ b/test/e2e/common.go
@@ -175,7 +175,7 @@ func EnsureImage(k8sVersion string) (imageURL string, imageChecksum string) {
 // DownloadFile will download a url and store it in local filepath.
 func DownloadFile(filePath string, url string) error {
        // Get the data
-       resp, err := http.Get(url) //nolint:noctx //#nosec
+       resp, err := http.Get(url) //nolint:noctx
        if err != nil {
                return err
        }
```

tuminoid commented:
> I was testing how to put this nolint and nosec together and I removed one but I didn't get any error
> 
> ```
> ./hack/gosec.sh
> + IS_CONTAINER=false
> + CONTAINER_RUNTIME=docker
> + GO_CONCURRENCY=8
> + '[' false '!=' false ']'
> + docker run --rm --env IS_CONTAINER=TRUE --volume /Users/kashif/Documents/Work/cluster-api-provider-metal3:/workdir:ro,z --entrypoint sh --workdir /workdir docker.io/securego/gosec:2.17.0@sha256:4ea9b6053eac43abda841af5885bbd31ee1bf7289675545b8858bcedb40b4fa8 /workdir/hack/gosec.sh
> + IS_CONTAINER=TRUE
> + CONTAINER_RUNTIME=podman
> + GO_CONCURRENCY=8
> + '[' TRUE '!=' false ]
> + export 'XDG_CACHE_HOME=/tmp/.cache'
> + gosec -severity medium --confidence medium -quiet -concurrency 8 '-exclude-dir=api' '-exclude-dir=test' ./...
> + cd api
> + gosec -severity medium --confidence medium -quiet -concurrency 8 ./...
> + cd test
> + gosec '-exclude=G107' -severity medium --confidence medium -quiet -concurrency 8 ./...
> kashif@Kashifs-MacBook-Pro cluster-api-provider-metal3 % git diff                       
> diff --git a/test/e2e/common.go b/test/e2e/common.go
> index 0c271d3a..9c9f218a 100644
> --- a/test/e2e/common.go
> +++ b/test/e2e/common.go
> @@ -175,7 +175,7 @@ func EnsureImage(k8sVersion string) (imageURL string, imageChecksum string) {
>  // DownloadFile will download a url and store it in local filepath.
>  func DownloadFile(filePath string, url string) error {
>         // Get the data
> -       resp, err := http.Get(url) //nolint:noctx //#nosec
> +       resp, err := http.Get(url) //nolint:noctx
>         if err != nil {
>                 return err
>         }
> ```

I added global exclude back to gosec.sh, but would like to get rid of it.

kashifest commented on test/e2e/common.go:
```diff
@@ -175,7 +175,7 @@ func EnsureImage(k8sVersion string) (imageURL string, imageChecksum string) {
 func DownloadFile(filePath string, url string) error {
 	// Get the data
-	resp, err := http.Get(url) //nolint:noctx
+	resp, err := http.Get(url) //nolint:noctx //#nosec
```
can we test
```suggestion
	resp, err := http.Get(url) //nolint:noctx,gosec 
```
with //#nosec we can also say if we want to skip specific rules only i.e.  `#nosec G401`

tuminoid replied:
Sometimes it wants `//#nosec` and sometimes `//nolint:gosec`... anyways, will be fixed. Thanks.

tuminoid commented:
Let's first fix the issues in ensure-golangci-lint here: https://github.com/metal3-io/cluster-api-provider-metal3/pull/1214 
I'll rebase on top of that and continue.

tuminoid commented:
```console
  Error: directive `//nolint:noctx,gosec` is unused for linter "gosec" (nolintlint)
  Error: directive `//nolint:gosec` is unused for linter "gosec" (nolintlint)
  Error: directive `//nolint:gosec` is unused for linter "gosec" (nolintlint)
  Error: directive `//nolint:gosec` is unused for linter "gosec" (nolintlint)
```

The action based linter is complaining about the `nolint:gosec` directives, while the `make lint` golangci-lint is happy. @kashifest  ideas?

tuminoid commented:
> ```
>   Error: directive `//nolint:noctx,gosec` is unused for linter "gosec" (nolintlint)
>   Error: directive `//nolint:gosec` is unused for linter "gosec" (nolintlint)
>   Error: directive `//nolint:gosec` is unused for linter "gosec" (nolintlint)
>   Error: directive `//nolint:gosec` is unused for linter "gosec" (nolintlint)
> ```
> 
> The action based linter is complaining about the `nolint:gosec` directives, while the `make lint` golangci-lint is happy. @kashifest ideas?

OK, it was complaining because they were ignored by `#nosec` already (on the previous line).

tuminoid commented:
/test-centos-e2e-integration-main
/test-centos-e2e-integration-main

tuminoid commented:
/test-centos-e2e-integration-main
/test-ubuntu-integration-main

tuminoid commented:
/test-ubuntu-integration-main

tuminoid commented:
/test-centos-e2e-integration-main
/test-ubuntu-integration-main

kashifest reviewed (commented):
/approve

tuminoid commented:
/test-centos-e2e-integration-main
/test-ubuntu-integration-main

lentzi90 reviewed (commented):
/lgtm
